apply-ParagraphStandards.py

```diff
@@ -0,0 +1,149 @@
+import glob
+import json
+import yaml
+import getopt, sys
+from pprint import pprint
+from dateutil import parser
+
+def usage():
+    print ('Usage:')
+    print ("\t -m                               : modify source files")
+    print ("\t -t:{test}, or --test {test}      : run specific tests, or All (default)")
+    print ("\t -h, or --help                    : display help")
+    print ("\t Test list:")
+    print ("\t\t All")
+    print ("\t\t MDBlockGeneration")
+    print ("\t\t CodeBlockFix")
+
+try:
+    opts, args = getopt.getopt(sys.argv[1:], "-t:hm", ["help", "test=", "--modify"])
+except getopt.GetoptError as err:
+    # print help information and exit:
+    print(err) # will print something like "option -a not recognized"
+    usage()
+    sys.exit(2)
```

Comment:
Let's not use `2` as an exit code as a matter of practice. Please use `1` instead. Although technically speaking, a user-defined script/program shouldn't use `1` either, it is a de facto standard when you don't need to distinguish various types of errors in your script. See http://tldp.org/LDP/abs/html/exitcodes.html.

Reply:
No, I suggested `1` instead of `2`. Zero (`0`) indicates success. If you follow the link I provided, you'll see that `2` is used for "Misuse of shell builtins (according to Bash documentation)". Further, unless you have something that calls this script and wants to be able to distinguish between various error conditions, there is no need to exit with any non-zero code other than `1`, which simply indicates "some" error occurred. It could also indicate a "falsey" condition, similar to how `test` and `grep` behave.

However, in cases where you do want to indicate different error conditions (rather than simply "some" error -- or "falsey" condition -- occurred), you will by all means want to use a value other than `1`, but such values should be in the range `3` to `125` (again, see the link).

Technically speaking, according to what I linked, `1` should not be used either, but it's use is so widely used and accepted, that it is really the de facto standard exit code to use to indicate that something (non-specific) went wrong or is "falsey".

Reply:
It was unclear to me what you were getting to on the use of 1.  

Now I do eventually for see a part of this script being used in a "lint" type test in a Jenkins job processing branches of the zeppelin-notebooks repo.  Because there are some features we cannot fix, yet could potentially be an issue, and we will want to see that in the Jenkins test when they notebooks are checked in.  In this case Jenkins will require the non-zero exit code to determine a failure.

Reply:
Not sure if you are saying you now understand my answer or that I have made things more confusing. 😄 

To summarize:

- use `0` to indicate success
- to indicate non-success when you don't need to distinguish one type of non-success from another, use `1`, which is the de facto general non-success code
- to indicate different types of "non-success" conditions, use different numbers in the range `[3, 125]`

Whichever you choose to use for non-success within a given script, do so consistently throughout the script. That is, if you don't need to distinguish different types of non-success, use `1` everywhere within the script where you want to indicate non-success.

Otherwise, choose a distinct value in the range `[3, 125]` for each distinct non-success condition you want to identify within the script. This would be necessary only if you want callers of your script to be able to distinguish between different non-success in order to execute appropriate logic based upon such non-success conditions, but this is not often the case, which is why `1` is so commonly used.

For example, `grep` uses the following:

- `0`: One or more lines were selected.
- `1`: No lines were selected.
- `>1`: An error occurred.

This makes sense since a caller of `grep` may wish to distinguish between these 3 conditions (2 of which are "non-success") and act appropriately. The choice of non-zero codes may be questionable based upon what I linked, but at least they are well-defined (specific to `grep`) and documented (in the `man` page), which is probably more important.

Conversely, if somebody is reading/debugging your script, seeing non-zero exit codes other than `1` could be confusing without documentation explaining such a choice, given what I've already described. That is, someone might wonder what specific condition your intending to indicate with your "non-success" code that is not `1`, when in fact your have no such intention other than a general "non-success" condition.

Wow. That was a lot of words. I better stop before I confuse myself as well, or even worse, begin to question my reasoning and have to start to elaborate some alternative.

Reply:
I got it after the second iteration.  I'm good with using 1.  The example code for python's getopt module includes the 2 on invalid arguments being passed in.  My thought process was only in that i want to indicate a non-zero for Jenkins to fail the job in that particular case.  Once we get to that point anyway.

I had already started on a second version of the script, the whole "standalone" docker was under some time pressure, so that was where it was at when it was needed.  I'll check in the new one shortly.